modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java

```diff
@@ -24,25 +24,35 @@
 import org.apache.ignite.internal.catalog.commands.AlterZoneParams;
 import org.apache.ignite.internal.catalog.commands.CreateHashIndexParams;
 import org.apache.ignite.internal.catalog.commands.CreateSortedIndexParams;
-import org.apache.ignite.internal.catalog.commands.CreateTableParams;
 import org.apache.ignite.internal.catalog.commands.CreateZoneParams;
 import org.apache.ignite.internal.catalog.commands.DropIndexParams;
 import org.apache.ignite.internal.catalog.commands.DropTableParams;
 import org.apache.ignite.internal.catalog.commands.DropZoneParams;
 import org.apache.ignite.internal.catalog.commands.RenameZoneParams;
+import org.apache.ignite.internal.catalog.commands.builders.CreateTableCommandBuilder;
 import org.apache.ignite.internal.manager.IgniteComponent;
 
 /**
  * The catalog manager provides schema manipulation methods and is responsible for managing distributed operations.
  */
 public interface CatalogManager extends IgniteComponent, CatalogService {
     /**
-     * Creates new table.
+     * Executes given command.
      *
-     * @param params Parameters.
-     * @return Operation future.
+     * <p>Accepts only those commands provided by builders returned by this very {@link CatalogManager}.
+     * Otherwise will throw {@link IllegalArgumentException}.
+     *
+     * @param command Command to execute.
+     * @return Future representing result of execution.
+     * @throws IllegalArgumentException If given command was created not by this manager.
+     * @see #createTableCommandBuilder()
+     */
+    CompletableFuture<Void> execute(CatalogCommand command) throws IllegalArgumentException;
+
+    /**
+     * Returns builder to create a command to create a new table.
      */
-    CompletableFuture<Void> createTable(CreateTableParams params);
+    CreateTableCommandBuilder createTableCommandBuilder();
```

Comment:
We use CatalogManager for creating command builder.
Then we use CatalogManager again to apply the command, and expect the CatalogCommand is  an command instance, which is created with the builder.

Looks. like we expose some internals to the outside.
As we don't need a CatalogManager to create the builder, then maybe this method could be moved to a separate Factory class?

Reply:
I agree that this looks rather strange. Besides, there is violation of Liskov Substitution Principle.

Considering only single command execution case, we actually may move method `execute` to the interface CatalogCommand, making a command self-contained. But I did not came up with a good solution to implement batched command. No matter what you do, you always need a guarantee that the command was created through the right builder.

One option is to introduce method `createBatch(List<Command> commands)` on a manager, another option is to introduce method `execute(List<CatalogCommand> commands)` on a manager. 

I decided to go with second option. With that said, having method `execute` on a manager for single command make api more consistent.

> maybe this method could be moved to a separate Factory class?

that's possible, but what will we do if we need access to some dynamic defaults from configuration, for instance? I think, static factory methods for such kind of objects not the best choice.